eventsourcing/src/test/java/org/axonframework/eventsourcing/eventstore/AnnotationBasedTagResolverTest.java

```diff
@@ -0,0 +1,521 @@
+            assertTrue(result.contains(new Tag("id", "123")));
+            assertTrue(result.contains(new Tag("customKey", "test")));
+        }
+
+        @Test
+        void shouldIgnoreTagsWithNullValues() {
+            // given
+            var payload = new NullableRecord("123", null);
+            var event = anEventMessage(payload);
+
+            // when
+            var result = tagResolver.resolve(event);
+
+            // then
+            var nullValueTagNotPresent = result.stream().noneMatch(it -> it.key().equals("nullValue"));
+            assertTrue(nullValueTagNotPresent);
+        }
+    }
+
+    @Nested
+    class ClassTests {
+
+        static class TestClass {
+
+            @EventTag
+            private final String id;
+
+            @EventTag(key = "customField")
+            private final Integer value;
+
+            private final String nonTagged;
+
+            TestClass(String id, Integer value, String nonTagged) {
+                this.id = id;
+                this.value = value;
+                this.nonTagged = nonTagged;
+            }
+
+            @EventTag(key = "taggedMethod")
+            public String getTaggedMethod() {
+                return "methodValue";
+            }
+
+            @EventTag(key = "customMethod")
+            public String getCustomTaggedMethod() {
+                return "customMethodValue";
+            }
+        }
+
+        @Test
+        void shouldResolveTagsFromFieldsAndMethods() {
+            // given
+            var payload = new TestClass("123", 456, "ignored");
+            var event = anEventMessage(payload);
+
+            // when
+            var result = tagResolver.resolve(event);
+
+            // then
+            assertEquals(4, result.size());
+            assertTrue(result.contains(new Tag("id", "123")));
+            assertTrue(result.contains(new Tag("customField", "456")));
+            assertTrue(result.contains(new Tag("taggedMethod", "methodValue")));
+            assertTrue(result.contains(new Tag("customMethod", "customMethodValue")));
+        }
+
+        @Test
+        void shouldNotResolveTagsFromNotAnnotatedMembers() {
+            // given
+            var payload = new TestClass("123", 456, "ignored");
+            var event = anEventMessage(payload);
+
+            // when
+            var result = tagResolver.resolve(event);
+
+            // then
+            assertFalse(result.contains(new Tag("nonTagged", "ignored")));
+        }
+    }
+
+    @Nested
+    class ComplexCases {
+
+        record ComplexRecord(
+                @EventTag String id,
+                @EventTag(key = "mapObject") Map<String, Integer> mapObject,
+                @EventTag(key = "listObject") List<Object> listObject,
+                @EventTag(key = "setObject") Set<Object> setObject,
+                @EventTag(key = "collectionObject") Collection<Object> collectionObject,
+                @EventTag(key = "iterableObject") Iterable<Object> iterableObject,
```

Comment:
What do we do with Collections / Iterable annotated? 
Question 1: Should we allow `@EventTag` annotation on them? If yes, do we treat is as value or do something like `iterable.map(v -> Tag.of(key,v)`? 
Question 2: We may allow it, but treat them as normal object (put as value) with `@EventTag` and introduce special annotation like `EventTags` whic hwill do `iterable.map(v -> Tag.of(key,v)`.

Reply:
Implemented here:
https://github.com/AxonFramework/AxonFramework/pull/3277/commits/b9e30867ee90babf4168f51a77b491aeaaa65af1